### PR TITLE
Clean up TaskBuilder tests.

### PR DESCRIPTION
**Pull request by jeschkies**
I split up the task builder tests into test suites. I hope the labels make some sense.

The tests have more of an BDD approach now:

```scala
"TaskBuilder" should {
  "given <offer> and <app definition>"
     val offer = ???
     val appDef = ???
     val task = builder....

     "define task" in { task should be('defined) }
     "set ports" in { ... }
}
```

The output becomes very nice:

```
.
.
.
TaskBuilder
  when given host ports 1001 and 1002
  - should set env variable PORT (0 milliseconds)
  - should set env variable PORT0 (0 milliseconds)
  - should set env variable PORT1 (0 milliseconds)
  - should not set env variable PORT_1 (1 millisecond)
  when given no ports
  - should not set any env variable (2 milliseconds)
  when given protocols
  - should set env variable PORT (0 milliseconds)
  - should set env variable PORT0 (0 milliseconds)
  - should set env variable PORT1 (0 milliseconds)
  - should set env variable PORT_HTTP (0 milliseconds)
  - should set env variable PORT_HTTPS (0 milliseconds)
.
.
.
```

Also, the next assertion are tested if one fails.

### CODE REVIEW

File: src/test/scala/mesosphere/mesos/TaskBuilderRolesTestSuite.scala [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+      }
+
+      "set the resource roles to marathon" in {
+        for (r <- taskInfo.getResourcesList.asScala) {
+          r.getRole should be("marathon")
+        }
+      }
+      // TODO(karsten): Do these numbers make sense?
+      "set an appropriate cpu share" in { resource("cpus") should be(ScalarResource("cpus", 1, "marathon")) }
+      "set an appropriate mem share" in { resource("mem") should be(ScalarResource("mem", 72, "marathon")) }
```
Comment (jeschkies) [inline src/test/scala/mesosphere/mesos/TaskBuilderRolesTestSuite.scala]: 72 seems odd to me. Does that make sense?

**Comment by jeschkies** [comment]
retest this please

**Comment by jeschkies** [comment]
retest this please

File: src/test/scala/mesosphere/mesos/TaskBuilderMesosContainerTestSuite.scala [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      "set the image docker" in { taskInfo.getContainer.getMesos.getImage.hasDocker should be (true) }
+      "set image docker credentials" in {
+        taskInfo.getContainer.getMesos.getImage.getDocker.hasCredential should be(true)
+        taskInfo.getContainer.getMesos.getImage.getDocker.getCredential.getPrincipal should be("aPrincipal")
+        taskInfo.getContainer.getMesos.getImage.getDocker.getCredential.hasSecret should be(true)
+        taskInfo.getContainer.getMesos.getImage.getDocker.getCredential.getSecret should be("aSecret")
+      }
+    }
+
+    /*
```
Comment (jdef) [inline src/test/scala/mesosphere/mesos/TaskBuilderMesosContainerTestSuite.scala]: why is this commented out?

File: src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala [hunk 3]
```diff
@@ -0,0 +1,840 @@
...
+      }
+      "add network-scope labels" in {
+        discoveryInfo.getPorts.getPorts(0).getLabels.getLabels(1).getKey should be("network-scope")
+        discoveryInfo.getPorts.getPorts(0).getLabels.getLabels(1).getValue should be("host")
+        discoveryInfo.getPorts.getPorts(1).getLabels.getLabels(1).getKey should be("network-scope")
+        discoveryInfo.getPorts.getPorts(1).getLabels.getLabels(1).getValue should be("container")
+      }
+    }
+
+    "given an offer with duplicated resources and an app definition" should {
```
Comment (jdef) [inline src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala]: how are "duplicated resources" relevant for this suite? what's really being tested here?

File: src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala [hunk 4]
```diff
@@ -0,0 +1,840 @@
...
+
+      val task: Option[(MesosProtos.TaskInfo, _)] = buildIfMatches(offer, appDef)
+      val Some((taskInfo, _)) = task
+      def resource(name: String): Resource = taskInfo.getResourcesList.asScala.find(_.getName == name).get
+      val portsResource: Resource = resource("ports")
+
+      "set an appropiate cpu share" in { resource("cpus") should be(ScalarResource("cpus", 1)) }
+      "set an appropiate mem share" in { resource("mem") should be(ScalarResource("mem", 64)) }
+      "set an appropiate disk share" in { resource("disk") should be(ScalarResource("disk", 1)) }
+      "???" in {
```
Comment (jdef) [inline src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala]: fully allocates the only two ports that exist in the available port range

File: src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala [hunk 5]
```diff
@@ -0,0 +1,840 @@
...
+        assert(portsResource.getRanges.getRangeList.asScala.map(range => range.getEnd - range.getBegin + 1).sum == 2)
+      }
+      "set unreserved ports resource role" in { portsResource.getRole should be(ResourceRole.Unreserved) }
+    }
+
+    // #1583 Do not pass zero disk resource shares to Mesos
```
Comment (jdef) [inline src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala]: @nfnt we should have learned our lesson here?

File: src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala [hunk 6]
```diff
@@ -0,0 +1,840 @@
...
+        acceptedResourceRoles = Some(Set(ResourceRole.Unreserved, "marathon"))
+      )
+      val Some((taskInfo, _)) = task
+      def resource(name: String): Resource = taskInfo.getResourcesList.asScala.find(_.getName == name).get
+      val portsResource: Resource = resource("ports")
+
+      "set an appropiate cpu share" in { resource("cpus") should be(ScalarResource("cpus", 1, "marathon")) }
+      "set an appropiate mem share" in { resource("mem") should be(ScalarResource("mem", 64, "marathon")) }
+      "set an appropiate disk share" in { resource("disk") should be(ScalarResource("disk", 1, "marathon")) }
+      "???" in {
```
Comment (jdef) [inline src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala]: ditto

File: src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala [hunk 7]
```diff
@@ -0,0 +1,840 @@
...
+        )
+
+      val task: Option[(MesosProtos.TaskInfo, _)] = buildIfMatches(offer, appDef)
+      val Some((taskInfo, _)) = task
+      def resourceOpt(name: String) = taskInfo.getResourcesList.asScala.find(_.getName == name)
+
+      "return a task with an empty disk resources" in { assert(resourceOpt("disk").isEmpty) }
+    }
+
+    "given an offer and an app definition to share resources" should {
```
Comment (jdef) [inline src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala]: maybe mention that role comes into play here?

File: src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala [hunk 8]
```diff
@@ -0,0 +1,840 @@
...
+
+      "return a serialized Mesos proto port definition with the correct protocols" in {
+        mesosPortDefinition.getProtocol should be("tcp,udp")
+      }
+      "return a serialized Mesos proto port definition with the correct port number" in {
+        mesosPortDefinition.getNumber should be(80)
+      }
+    }
+
+    "given an offer and an app definition with port mapping and host port" should {
```
Comment (jdef) [inline src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala]: "... with a container-port-only mapping"?

Comment (jdef) [inline src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala]: "... when using virtual networking"

File: src/test/scala/mesosphere/mesos/TaskBuilderEnvironmentTestSuite.scala [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+
+      "set env variable PORT" in { env("PORT") should be("1001") }
+      "set env variable PORT0" in { env("PORT0") should be("1001") }
+      "set env variable PORT1" in { env("PORT1") should be("1002") }
+
+      "set env variable PORT_80" in { env("PORT_80") should be("1001") }
+      "set env variable PORT_8080" in { env("PORT_8080") should be("1002") }
+    }
+
+    "given deckared named ports" should {
```
Comment (jeschkies) [inline src/test/scala/mesosphere/mesos/TaskBuilderEnvironmentTestSuite.scala]: Fixed

File: src/test/scala/mesosphere/mesos/TaskBuilderEnvironmentTestSuite.scala [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+      val env = TaskBuilder.taskContextEnv(runSpec = runSpec, Some(Task.Id("taskId")))
+        .filterKeys(_.startsWith("MARATHON_APP_LABEL"))
+
+      "set env variable MARATHON_APP_LABELS" in { env("MARATHON_APP_LABELS") should be("OTHER_LABEL_A LABEL LABEL_WITH_INVALID_CHARS") }
+      "set env variable MARATHON_APP_LABEL_LABEL" in { env("MARATHON_APP_LABEL_LABEL") should be("VALUE1") }
+      "set env variable MARATHON_APP_LABEL_LABEL_WITH_INVALID_CHARS" in { env("MARATHON_APP_LABEL_LABEL_WITH_INVALID_CHARS") should be("VALUE2") }
+      "set env variable MARATHON_APP_LABEL_OTHER_LABEL_A" in { env("MARATHON_APP_LABEL_OTHER_LABEL_A") should be("VALUE3") }
+    }
+
+    "given an app definition with an DOCKER containter image" should {
```
Comment (jeschkies) [inline src/test/scala/mesosphere/mesos/TaskBuilderEnvironmentTestSuite.scala]: I'm not sure. I just ported the tests we had.

File: src/test/scala/mesosphere/mesos/TaskBuilderMesosContainerTestSuite.scala (hunk 2 above)
Comment (jeschkies) [inline src/test/scala/mesosphere/mesos/TaskBuilderMesosContainerTestSuite.scala]: I triaged a compiling error. I've forgot to enable this test again.

File: src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala (hunk 5 above)
Comment (nfnt) [inline src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala]: Yeah, we should have 😄

File: src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala (hunk 3 above)
Comment (jeschkies) [inline src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala]: As far as I can see the only difference to other tests is that we add more resources.

File: src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala (hunk 7 above)
Comment (jeschkies) [inline src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala]: done

File: src/test/scala/mesosphere/mesos/TaskBuilderEnvironmentTestSuite.scala [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+      "set env variable PORT_8081" in { env("PORT_8081") should be("1001") }
+      "set env variable PORT_HTTP" in { env("PORT_HTTP") should be("1000") }
+      "set env variable PORT_JABBER" in { env("PORT_JABBER") should be("1001") }
+    }
+
+    "given a task command with ports and ports mappings" should {
+      val command =
+        TaskBuilder.commandInfo(
+          runSpec = AppDefinition(
+            portDefinitions = PortDefinitions(22, 23),
```
Comment (jeschkies) [inline src/test/scala/mesosphere/mesos/TaskBuilderEnvironmentTestSuite.scala]: According to the original tests, yes. I'm not sure if this is a bug then.

**Comment by jeschkies** [comment]
retest this please

**Comment by jeschkies** [comment]
restest this please

**Comment by jeschkies** [comment]
retest this please

**Comment by jeschkies** [comment]
retest this please

**Comment by jeschkies** [comment]
@jdef I'll update until it's reviewed ;)

**Comment by jeschkies** [comment]
retest this please

**Comment by jeschkies** [comment]
Abandon refactor cause of pods merge.

**Pull request closed without merging**